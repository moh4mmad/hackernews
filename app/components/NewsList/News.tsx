'use client';
import React, { useEffect, useRef } from 'react';
import { Story } from '../../types/story';

export default function News({
  story,
  index,
  isLastStory,
  fetchNextStories,
}: {
  story: Story;
  index: number;
  isLastStory: boolean;
  fetchNextStories: () => void;
}) {
  const newsRef = useRef(null);

  // To execute fetchNextStories function when the last story is on the screen
  useEffect(() => {
    if (!newsRef?.current || !isLastStory) return;

    const observer = new IntersectionObserver(([entry]) => {
      if (entry.isIntersecting) {
        fetchNextStories();
        observer.unobserve(entry.target);
      }
    });
    observer.observe(newsRef.current);
  }, [isLastStory]);

  return (
    <div>
      <div className="animate-tilt absolute -inset-0.5 rounded-lg bg-gradient-to-r from-pink-600 to-purple-600 opacity-75 blur transition duration-1000 group-hover:opacity-100 group-hover:duration-200"></div>
      <a
        href={story?.url}
        target="_blank"
        rel="noreferrer"
        ref={newsRef}
        className="relative flex items-center divide-x divide-gray-600 rounded-lg bg-black px-7 py-3 leading-none"
      >
        <span className="flex items-center space-x-5">
          <big className="text-3xl font-bold text-pink-600">{index + 1}</big>
          <span className="py-3 pr-6 font-bold text-gray-100">
            {story.title}
          </span>
        </span>
      </a>
    </div>
  );
}
