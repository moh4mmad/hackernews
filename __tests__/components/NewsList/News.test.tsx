import { render, screen } from '@testing-library/react';
import News from '@/app/components/NewsList/News';

describe('Page News Component', () => {
  test('should render News component', () => {
    const content = render(<News />);

    const linkElement = screen.getByRole('link', {
      name: /1 Show HN: DriftDB is an open source WebSocket backend for real-time apps/i,
    });
    expect(linkElement).toBeInTheDocument();
    expect(linkElement).toHaveAttribute('href', 'https://driftdb.com/');
    expect(linkElement).toHaveAttribute('target', '_blank');
    expect(content).toMatchSnapshot();
  });
});
