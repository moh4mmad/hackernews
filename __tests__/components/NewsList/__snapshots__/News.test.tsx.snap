// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page News Component should render News component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="animate-tilt absolute -inset-0.5 rounded-lg bg-gradient-to-r from-pink-600 to-purple-600 opacity-75 blur transition duration-1000 group-hover:opacity-100 group-hover:duration-200"
        />
        <a
          class="relative flex items-center divide-x divide-gray-600 rounded-lg bg-black px-7 py-3 leading-none"
          href="https://driftdb.com/"
          rel="noreferrer"
          target="_blank"
        >
          <span
            class="flex items-center space-x-5"
          >
            <big
              class="text-3xl font-bold text-pink-600"
            >
              1
            </big>
            <span
              class="py-3 pr-6 font-bold text-gray-100"
            >
              Show HN: DriftDB is an open source WebSocket backend for real-time apps
            </span>
          </span>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="animate-tilt absolute -inset-0.5 rounded-lg bg-gradient-to-r from-pink-600 to-purple-600 opacity-75 blur transition duration-1000 group-hover:opacity-100 group-hover:duration-200"
      />
      <a
        class="relative flex items-center divide-x divide-gray-600 rounded-lg bg-black px-7 py-3 leading-none"
        href="https://driftdb.com/"
        rel="noreferrer"
        target="_blank"
      >
        <span
          class="flex items-center space-x-5"
        >
          <big
            class="text-3xl font-bold text-pink-600"
          >
            1
          </big>
          <span
            class="py-3 pr-6 font-bold text-gray-100"
          >
            Show HN: DriftDB is an open source WebSocket backend for real-time apps
          </span>
        </span>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
